File: 剑指offer刷题/剑指Offer-JavaScript实现/23_VerifySquenceOfBST.js
// 23.二叉搜索树的后序遍历序列

// 输入一个整数数组，判断该数组是不是某二叉搜索树的后序遍历的结果。如果是则输出Yes,否则输出No。假设输入的数组的任意两个数字都互不相同。

function VerifySquenceOfBST(sequence)
{
    // write code here
    if(sequence.length <= 0){
        return false
    }
    if(sequence.length == 1){
        return true
    }
    let length = sequence.length
    let root = sequence[length-1]
    let i = 0
    while(sequence[i]<root){
        i += 1 
    }
    let sepIndex = i
    for(let j = i; j<length; j++){
        if(sequence[j]<root){
            return false
        }
    }
    let leftSquence = sequence.slice(0,sepIndex-1)
    let rightSquence = sequence.slice(sepIndex+1,length-1)
    let left = true
    let right = true
    if(leftSquence.length>0){
        left = VerifySquenceOfBST(leftSquence)
    }    
    if(rightSquence.length>0){
        right = VerifySquenceOfBST(rightSquence)
    }  
    return left && right
}

File: SwordToOffer/SwordToOffer-PythonSolution/23_VerifySquenceOfBST.py
# 23.二叉搜索树的后序遍历序列

# 输入一个整数数组，判断该数组是不是某二叉搜索树的后序遍历的结果。如果是则输出Yes,否则输出No。假设输入的数组的任意两个数字都互不相同。

# -*- coding:utf-8 -*-
class Solution:
    def VerifySquenceOfBST(self, sequence):
        # write code here
        length = len(sequence)
        if(length<=0):
            return False
        if(length==1):
            return True
        i = 0
        root = sequence[-1]
        while(sequence[i]<root):
            i += 1
        sepIndex = i
        for j in range(i,length):
            if(sequence[j]<root):
                return False
        leftSquence = sequence[0:sepIndex]
        rightSquence = sequence[sepIndex:-2]
        left,right = True, True
        if(len(leftSquence)>0):
            left = self.VerifySquenceOfBST(leftSquence) 
        if(len(rightSquence)<0):
            rigth = self.VerifySquenceOfBST(rightSquence) 
        return left and right